contracts/utils/convex/ConvexStakingWrapperYield.sol

```diff
@@ -0,0 +1,111 @@
+//https://etherscan.io/address/0x3ba207c25a278524e1cc7faaea950753049072a4#code
+// SPDX-License-Identifier: BUSL-1.1
+pragma solidity 0.8.6;
+pragma experimental ABIEncoderV2;
+
+import './ConvexStakingWrapper.sol';
+import "@yield-protocol/utils-v2/contracts/access/AccessControl.sol";
+
+struct Balances {
+    uint128 art; // Debt amount
+    uint128 ink; // Collateral amount
+}
+
+interface ICauldron {
+    /// @dev Each vault records debt and collateral balances_.
+    function balances(bytes12 vault) external view returns (Balances memory);
+}
+
+//Staking wrapper for Yield platform
+//use convex LP positions as collateral while still receiving rewards
+contract ConvexStakingWrapperYield is ConvexStakingWrapper,AccessControl {
+    using SafeERC20 for IERC20;
+    using Address for address;
+
+
+    mapping(address=>bytes12[]) public vaults;
+    address cauldron;
+
+    constructor() public {}
+
+    function initialize(
+        address _curveToken,
+        address _convexToken,
+        address _convexPool,
+        uint256 _poolId,
+        address _vault
+    ) external override auth{
+        require(!isInit, 'already init');
+        owner = address(0xa3C5A1e09150B75ff251c1a7815A07182c3de2FB); //TODO: Find why this needs to be set to convex multisig
+        emit OwnershipTransferred(address(0), owner);//TODO: Find why this needs to be done
+        _tokenname = string(abi.encodePacked('Staked ', ERC20(_convexToken).name(), ' Yield'));
+        _tokensymbol = string(abi.encodePacked('stk', ERC20(_convexToken).symbol(), '-yield'));
+        isShutdown = false;
+        isInit = true;
+        curveToken = _curveToken;
+        convexToken = _convexToken;
+        convexPool = _convexPool;
+        convexPoolId = _poolId;
+        collateralVault = _vault; //TODO: Add the join address
+
+        //add rewards
+        addRewards();
+        setApprovals();
+    }
+
+    function setCauldron(address _cauldron) external auth{
+        require(_cauldron!=address(0), 'cauldron address cannot be 0');
+        cauldron = _cauldron;
+    }
+
+    // Set the locations of vaults where the user's funds have been deposited & the accounting is kept
+    function setVault(address _account, bytes12 _vault) external auth {
```

Comment:
Instead of an `auth` function, you can remove the `account` parameter, and read the owner from the cauldron using `cauldron.vaults(vaultId).owner`. Anyone can register anyone's vault, which is fine.

Then, we can have a Ladle module that takes a vaultId as parameter and calls this function, so that we can embed `setVault` in a Ladle batch.

It would be a module, because the Ladle keeps in `cachedVaultId` the vaultId of a vault that would just have been created in the same batch. If `convexModule.setVault` receives a zero as the vaultId, it calls `convexStakingWrapperYield.setVault(cachedVaultId)`, otherwise it calls `convexStakingWrapperYield.setVault(vaultId)`.

Reply:
Should we also handle the wrapping & unwrapping of the convex token via this module? This would allow the user to deposit the convex token and we could take care of the wrapping & unwrapping for them.

Reply:
My last comment could be ignored as we can directly add the wrapper as an integration and handle it through that.